bob-access/src/credentials.rs

```diff
@@ -9,18 +9,71 @@ use tonic::Request;
 
 use crate::{error::Error, extractor::ExtractorExt, Authenticator};
 
-#[derive(Debug, Default, Clone)]
-pub struct Credentials {
+#[derive(Debug, Default)]
+pub struct RequestCredentials {
     address: Option<SocketAddr>,
     kind: Option<CredentialsKind>,
 }
 
+impl RequestCredentials {
+    pub fn builder() -> RequestCredentialsBuilder {
+        RequestCredentialsBuilder::default()
+    }
+
+    pub fn kind(&self) -> Option<&CredentialsKind> {
+        self.kind.as_ref()
+    }
+
+    pub fn ip(&self) -> Option<SocketAddr> {
+        self.address
+    }
+}
+
+const RESOLVE_THRESHOLD: u8 = 3;
+
+#[derive(Debug, Clone)]
+enum ResolveState {
+    Waiting(u8),
+    Pending,
+}
+
+impl Default for ResolveState {
+    fn default() -> Self {
+        ResolveState::Waiting(0)
+    }
+}
+
+impl ResolveState {
+    fn needs_resolve(&mut self, unresolved: bool) -> bool {
+        match self {
+            ResolveState::Waiting(n) => {
+                if unresolved {
+                    *n += 1;
```

Comment:
In this form it has a risk of integer overflow with the following panic

Reply:
You still should use `wrapping_add` to avoid panics in rare corner case (when even u32 overflows)